### DISCUSSION TITLE
Possible OpenDDS resource leak when participant disconnects/reconnects due to network interruption

### DISCUSSION THREAD

**Discussion by smnrgrs** [OpenDDS/OpenDDS#4291] [Q&A] [answered]
We have an OpenDDS (3.22.0) based system running on Windows which uses a mixture of reliable and best effort topics, with RTPS UDP discovery/transport. It is in use on a network with an unreliable link between some of the participants.

We have about 10 PCs, each of which runs about 10 processes which are DDS participants.

We have been investigating a memory leak which seems to be triggered by the network disconnecting/reconnecting between participants. In an artificial test where the same participant PC is connected/disconnected repeatedly (2 min on network, 2min off) we see memory growth of the order 10M/Hr.

We're trying to distil this down to a simple demonstration that reproduces the problem, but haven't got there yet.

Would the team have any suggestions for areas of software we could monitor/debug to try and get to the bottom of this - perhaps resources allocated for a participant on discovery that never get cleaned up if the participant disappears from the network.

Thanks

**Comment by jrw972** [comment]
If your experiment doesn't actually send samples and only performs discovery, then there is a high probability that you are correct about discovery being the problem.

This PR that was after 3.25 comes to mind https://github.com/OpenDDS/OpenDDS/pull/4216

**Reply by smnrgrs** [reply]
Thanks for the suggestion. I am compiling in the patch to see if it makes any difference.

**Comment by jrw972** [comment]
Some other questions that come to mind:

1. Are you using security?  Security is largely driven by discovery so, like discovery, a data structure in security might be growing without bound.  Turning on `DCPS::security_debug.bookkeeping` should log the sizes of these data structures.
2. Are you using fragmentation?  Fragments have to be stored for a little while to rebuild the sample.  It is possible that this caching mechanism is broken and/or holding onto fragments longer than you want.  This is governed by `fragment_reassembly_timeout`.  I don't think you can configure this for RtpsDiscovery but you can for a normal transport.
3. What is the lease time for discovery?  If using the default of 5 minutes, then 2 minutes on and then 2 minutes off shouldn't really cause rediscovery.
4. Are the lease times asymmetric?  Asymmetric lease times create an interesting problem where one participant wants to go through discovery while the other does not.  There may be a bug in the code that handles this.  Even symmetric lease times can trigger this depending on timing.

**Reply by smnrgrs** [reply]
Is there a way of switching on security_debug.bookkeeping from the rtps.ini file, along the lines of DCPSTransportDebugLevel=0
DCPSDebugLevel=0?

**Reply by jrw972** [reply]
```
DCPSecurityDebug=1
DCPSSecurityDebugLevel=4
```

https://opendds.readthedocs.io/en/latest-release/devguide/run_time_configuration.html#security-debug-logging

**Comment by smnrgrs** [comment, marked as answer]
We have been memory profiling, and discovered that built in topic data is the source of the leak. We simplified our test to one where a participant repeatedly subscribed to topics and then shut down. This causes several megabytes per hour to be used in other running participants. 

Through debugging the OpenDDS libraries we found that the bit_autopurge_disposed_samples_delay and bit_autopurge_nowriter_samples_delay QoS values default to infinity. This has the effect of never releasing the memory allocated for the built in topic data received about the shut down/disconnected subscribers. 

Setting these:
TheServiceParticipant->bit_autopurge_disposed_samples_delay(duration);
TheServiceParticipant->bit_autopurge_nowriter_samples_delay(duration);
and applying the patch from https://github.com/OpenDDS/OpenDDS/pull/4216 (a smaller leak) seems to fix our issue.

Is there a reason for the bit_autopurge delays defaulting to infinity? It seems to cause OpenDDS to leak resources by default.

**Reply by jrw972** [reply]
I had to deal with the same issue in the RtpsRelay.  https://github.com/OpenDDS/OpenDDS/pull/1940
I apologize for not remembering that as it could have saved you a lot of effort.  

@mitza-oci is right that it's in the spec so we are probably not going to change it.
However, what can we do to prevent future users from stumbling on this again?
Things that come to mind are

1. A warning in the log when they are set to infinity.
2. A warning in the DevGuide.
    What is the appropriate place for the warning in the DevGuide?
    I don't think that https://opendds.readthedocs.io/en/latest-release/devguide/built_in_topics.html is adequate since this issue will affect users by just having the BITs enabled.
    My preference would be a dedicated "Gotchas" section.
    @smnrgrs As a user, what is your preference?

**Reply by smnrgrs** [reply]
It would be useful to cover bit_autopurge_disposed_samples_delay and bit_autopurge_nowriter_samples_delay in the Quality of Service chapter/section. Autopurge is covered for readers and writers, but not for built in topics. As users of OpenDDS we're used to referring to the QoS section pretty often.

We think we're seeing the instances_ map in DataReaderImpl.cpp growing out of control too. Is there another QoS we've missed that might help with that?

Specifically, DCPSPublication, DCPSSubscription, DCPSParticipant and OpenDDSParticipantLocation topics are accumulating handles in the instances_ map relating to participants which are shut down/disconnected. 

Thanks

**Reply by jrw972** [reply]
Thanks for the feedback.

I ran an experiment on the master branch by setting the `bit_autopurge_*` to 1 second, the participant resend period to 2 seconds, the participant lease duration to 1 minute, starting a pair of participants, and then killing one to see if the instance is removed from the map.  I didn't see a significant difference from 3.22 but I may be looking in the wrong place.

In any case, I used a snippet like this:
```
void
DataReaderImpl::release_instance(DDS::InstanceHandle_t handle)
{
  ACE_DEBUG((LM_DEBUG, "### releasing instance %d on topic %C\n", handle, topic_servant_->topic_name()));
```

In my case, the instances for the BITs were removed.
If a similar experiment doesn't work for you, then the issue is probably in SEDP.
When doing your experiment, remember that you won't see `release_instance` until the lease duration + resend period + `bit_autopurge_*`.
If these are set large enough, then I would expect to see the number of instances saturate at some level and stay constant once incoming=outgoing.

**Reply by smnrgrs** [reply]
Just to be clear, is it the case that if a particular reader/writer QoS lease duration is not set (and therefore defaults to infinity) then the built in topic data associated with those readers/writers will also have a lease duration of infinity?

**Reply by jrw972** [reply]
To clarify, I was referring to the lease duration for participants in RTPS discovery and not the lease duration for liveliness.  Even the lease duration for the BIT readers is not relevant.

A participant should be removed from discovery and the BITs when a participant doesn't receive an SPDP message for their lease duration.

**Reply by smnrgrs** [reply]
If we have a participant which repeatedly (because it runs as a service) crashes/restarts at start-up, other participants are growing in memory. MTuner suggests the used memory is used in the following call stack:
OpenDDS::RTPS::Spdp::SpdpTransport::handle_input(void * ptr64)	
void OpenDDS::RTPS::Spdp::handle_participant_data
void OpenDDS::RTPS::Sedp::associate(struct OpenDDS::RTPS::DiscoveredParticipant ...
void OpenDDS::RTPS::Sedp::process_association_records_i
void OpenDDS::RTPS::Sedp::Writer::transport_assoc_done(int,struct OpenDDS::DCPS::GUID_t const & ptr64)	Sedp.cpp	3212	c:\dev\OpenDDS\2019_X64\debug\dds\DCPS\RTPS\Sedp.cpp
void OpenDDS::RTPS::Sedp::write_durable_publication_data(struct OpenDDS::DCPS::GUID_t const & ptr64,bool)	Sedp.cpp	4649	c:\dev\OpenDDS\2019_X64\debug\dds\DCPS\RTPS\Sedp.cpp
OpenDDS::RTPS::Sedp::Writer::write_parameter_list(class OpenDDS::RTPS::ParameterList const & ptr64,struct OpenDDS::DCPS::GUID_t const & ptr64,class OpenDDS::DCPS::SequenceNumber & ptr64)
void OpenDDS::DCPS::DataLink::send(class OpenDDS::DCPS::TransportQueueElement * ptr64)	DataLink.inl	102	c:\dev\OpenDDS\2019_X64\debug\dds\DCPS\transport\framework\DataLink.inl

We see the debug line **RtpsUdpDataLink::customize_queue_element() - storing durable data for...** every time the broken participant starts up. 

Of course we should fix/stop the broken participant. But is there a way to make the data above get cleaned up, maybe a timeout or QoS?
Thanks

**Reply by jrw972** [reply]
If the `LeaseDuration` (https://opendds.readthedocs.io/en/latest-release/devguide/run_time_configuration.html) for the restarting participant is the default of 5 minutes, the memory should stop growing after 5 minutes.  The amount of memory is proportional to the number of restarts in the `LeaseDuration` multiplied by the number of publications and subscriptions in the non-restarting participant.  Thus, reducing the `LeaseDuration` for the restarting participant will limit the amount of memory growth.

The RtpsRelay has a featured called `-RestartDetection` (https://opendds.readthedocs.io/en/latest-release/devguide/internet_enabled_rtps.html) which addresses this problem by looking at the remote IP address and first bytes of the GUID to determine if a remote participant has restarted.  This is done at the application level and you are more than welcome to adapt it for your use.  Alternatively, it might be a good candidate for incorporating it as a feature in RtpsDiscovery.

**Comment by mitza-oci** [comment]
> Is there a reason for the bit_autopurge delays defaulting to infinity?

It's in the DDS spec

> It seems to cause OpenDDS to leak resources by default.

Memory that's reachable is not leaked

**Reply by jwillemsen** [reply]
Something to report on the OMG DDS spec?
